File: .llm/runs/claude-harness-profile-rfc-benchmark-shzhgv--polyglot-protocol-rfc/bench/spikes/k1/emit-frames.go
// K1 emitter (Go) — interleaves an adversarial log corpus with sentinel-framed protocol
// frames from concurrent goroutines. Frames are single write() calls <= 4096 bytes (the
// PIPE_BUF atomicity rule the protocol will mandate); logs are deliberately hostile.
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"
)

const SENTINEL = "\x00NSF\x00"

func main() {
	nLogs := 10000
	nFrames := 200
	var wg sync.WaitGroup
	var mu sync.Mutex // logs use a mutex only sometimes — half the writers are unsynchronized

	// frame writer goroutine: unsynchronized single-write frames (atomicity via PIPE_BUF)
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < nFrames; i++ {
			payload := strings.Repeat("x", 64*(i%16))
			h := sha256.Sum256([]byte(payload))
			frame := map[string]any{
				"v": 1, "t": "progress", "seq": i,
				"payload": payload, "sha": hex.EncodeToString(h[:8]),
			}
			b, _ := json.Marshal(frame)
			line := SENTINEL + string(b) + "\n"
			if len(line) > 4096 {
				panic("frame exceeds PIPE_BUF budget")
			}
			os.Stdout.WriteString(line) // one write syscall
		}
	}()

	// hostile log writers
	for w := 0; w < 4; w++ {
		wg.Add(1)
		go func(w int) {
			defer wg.Done()
			for i := 0; i < nLogs/4; i++ {
				var line string
				switch i % 8 {
				case 0:
					line = fmt.Sprintf("worker %d: plain log line %d\n", w, i)
				case 1:
					line = fmt.Sprintf("{\"looks\":\"like json\",\"seq\":%d,\"success\":true}\n", i) // D-3 trap
				case 2:
					line = SENTINEL + "this is not valid json {{{\n" // sentinel-lookalike
				case 3:
					line = fmt.Sprintf("mid-line %s{\"t\":\"result\"} embedded sentinel %d\n", SENTINEL, i)
				case 4:
					line = strings.Repeat("A", 1024*1024) + "\n" // 1MB line
				case 5:
					line = string([]byte{0xff, 0xfe, 0x80, 0x81}) + " binary-ish\n" // invalid UTF-8
				case 6:
					line = "\r\n" // CRLF/empty
				default:
					line = fmt.Sprintf("{\"t\":\"progress\",\"seq\":%d}\n", i) // frame-shaped, NO sentinel
				}
				if i%2 == 0 {
					mu.Lock()
					os.Stdout.WriteString(line)
					mu.Unlock()
				} else {
					os.Stdout.WriteString(line) // unsynchronized: big lines may shred
				}
			}
		}(w)
	}
	wg.Wait()
	// terminal result frame
	res := map[string]any{"v": 1, "t": "result", "seq": nFrames, "outcome": "ok", "framesEmitted": nFrames}
	b, _ := json.Marshal(res)
	os.Stdout.WriteString(SENTINEL + string(b) + "\n")
}

File: .llm/runs/claude-harness-profile-rfc-benchmark-shzhgv--polyglot-protocol-rfc/bench/spikes/k1/emit_frames.py
# K1 emitter (python3) — same adversarial contract as the Go emitter, via threads.
# Frames: single os.write() <= 4096 bytes (PIPE_BUF atomicity). Logs: hostile.
import hashlib
import json
import os
import sys
import threading

SENTINEL = b"\x00NSF\x00"
N_LOGS = 10000
N_FRAMES = 200
mu = threading.Lock()
fd = sys.stdout.fileno()


def frames() -> None:
    for i in range(N_FRAMES):
        payload = "x" * (64 * (i % 16))
        sha = hashlib.sha256(payload.encode()).hexdigest()[:16]
        frame = {"v": 1, "t": "progress", "seq": i, "payload": payload, "sha": sha}
        line = SENTINEL + json.dumps(frame, separators=(",", ":")).encode() + b"\n"
        assert len(line) <= 4096, "frame exceeds PIPE_BUF budget"
        os.write(fd, line)  # one write syscall


def logs(w: int) -> None:
    for i in range(N_LOGS // 4):
        k = i % 8
        if k == 0:
            line = f"worker {w}: plain log line {i}\n".encode()
        elif k == 1:
            line = json.dumps({"looks": "like json", "seq": i, "success": True}).encode() + b"\n"
        elif k == 2:
            line = SENTINEL + b"this is not valid json {{{\n"
        elif k == 3:
            line = b"mid-line " + SENTINEL + b'{"t":"result"} embedded %d\n' % i
        elif k == 4:
            line = b"A" * (1024 * 1024) + b"\n"
        elif k == 5:
            line = bytes([0xFF, 0xFE, 0x80, 0x81]) + b" binary-ish\n"
        elif k == 6:
            line = b"\r\n"
        else:
            line = json.dumps({"t": "progress", "seq": i}).encode() + b"\n"
        if i % 2 == 0:
            with mu:
                os.write(fd, line)
        else:
            os.write(fd, line)


threads = [threading.Thread(target=frames)] + [
    threading.Thread(target=logs, args=(w,)) for w in range(4)
]
for t in threads:
    t.start()
for t in threads:
    t.join()
res = {"v": 1, "t": "result", "seq": N_FRAMES, "outcome": "ok", "framesEmitted": N_FRAMES}
os.write(fd, SENTINEL + json.dumps(res, separators=(",", ":")).encode() + b"\n")
